calibrate.py

```diff
@@ -0,0 +1,187 @@
+gridspace_x = np.linspace(workspace_limits[0][0], workspace_limits[0][1], 1 + (workspace_limits[0][1] - workspace_limits[0][0])/calib_grid_step)
+gridspace_y = np.linspace(workspace_limits[1][0], workspace_limits[1][1], 1 + (workspace_limits[1][1] - workspace_limits[1][0])/calib_grid_step)
+gridspace_z = np.linspace(workspace_limits[2][0], workspace_limits[2][1], 1 + (workspace_limits[2][1] - workspace_limits[2][0])/calib_grid_step)
+calib_grid_x, calib_grid_y, calib_grid_z = np.meshgrid(gridspace_x, gridspace_y, gridspace_z)
+num_calib_grid_pts = calib_grid_x.shape[0]*calib_grid_x.shape[1]*calib_grid_x.shape[2]
+calib_grid_x.shape = (num_calib_grid_pts,1)
+calib_grid_y.shape = (num_calib_grid_pts,1)
+calib_grid_z.shape = (num_calib_grid_pts,1)
+calib_grid_pts = np.concatenate((calib_grid_x, calib_grid_y, calib_grid_z), axis=1)
+
+measured_pts = []
+observed_pts = []
+observed_pix = []
+
+# Move robot to home pose
+print('Connecting to robot...')
+robot = Robot(False, None, None, workspace_limits,
+              tcp_host_ip, tcp_port, rtc_host_ip, rtc_port,
+              False, None, None)
+robot.open_gripper()
+
+# Slow down robot
+robot.joint_acc = 1.4
+robot.joint_vel = 1.05
+
+# Make robot gripper point upwards
+robot.move_joints([-np.pi, -np.pi/2, np.pi/2, 0, np.pi/2, np.pi])
+
+# Move robot to each calibration point in workspace
+print('Collecting data...')
+for calib_pt_idx in range(num_calib_grid_pts):
+    tool_position = calib_grid_pts[calib_pt_idx,:]
+    robot.move_to(tool_position, tool_orientation)
+    time.sleep(1)
+    
+    # Find checkerboard center
+    checkerboard_size = (3,3)
+    refine_criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 30, 0.001)
+    camera_color_img, camera_depth_img = robot.get_camera_data()
+    bgr_color_data = cv2.cvtColor(camera_color_img, cv2.COLOR_RGB2BGR)
+    gray_data = cv2.cvtColor(bgr_color_data, cv2.COLOR_RGB2GRAY)
+    checkerboard_found, corners = cv2.findChessboardCorners(gray_data, checkerboard_size, None, cv2.CALIB_CB_ADAPTIVE_THRESH)
+    if checkerboard_found:
+        corners_refined = cv2.cornerSubPix(gray_data, corners, (3,3), (-1,-1), refine_criteria)
+
+        # Get observed checkerboard center 3D point in camera space
+        checkerboard_pix = np.round(corners_refined[4,0,:]).astype(int)
+        checkerboard_z = camera_depth_img[checkerboard_pix[1]][checkerboard_pix[0]]
+        checkerboard_x = np.multiply(checkerboard_pix[0]-robot.cam_intrinsics[0][2],checkerboard_z/robot.cam_intrinsics[0][0])
+        checkerboard_y = np.multiply(checkerboard_pix[1]-robot.cam_intrinsics[1][2],checkerboard_z/robot.cam_intrinsics[1][1])
+        if checkerboard_z == 0:
+            continue
+
+        # Save calibration point and observed checkerboard center
+        observed_pts.append([checkerboard_x,checkerboard_y,checkerboard_z])
+        # tool_position[2] += checkerboard_offset_from_tool
+        tool_position = tool_position + checkerboard_offset_from_tool
```

Comment:
What is tool_position, or more specifically, what is the tool they are referring to?

Reply:
I believe the tool_position is referring to the gripper. The code sets the tool_position to 'calib_grid_pts' that does calibration of the camera/gripper. Part of the paper: " The camera is localized with respect to the robot base by an automatic calibration procedure, during which the camera tracks the location of a checkerboard pattern taped onto the gripper. The calibration optimizes for extrinsics as the robot moves the gripper over a grid of 3D locations (predefined with respect to robot coordinates) within the camera field of view"